platform/ios/app/MBXFrameTimeGraphView.m

```diff
@@ -0,0 +1,127 @@
+        [self commonInit];
+    }
+    return self;
+}
+
+- (void)commonInit {
+    self.userInteractionEnabled = NO;
+    self.layer.opacity = 0.9f;
+
+    self.scrollLayer = [CAScrollLayer layer];
+    self.scrollLayer.scrollMode = kCAScrollHorizontally;
+    self.scrollLayer.masksToBounds = YES;
+    [self.layer addSublayer:self.scrollLayer];
+
+    self.thresholdLayer = [CAShapeLayer layer];
+    self.thresholdLayer.fillColor = [UIColor darkGrayColor].CGColor;
+    [self.layer addSublayer:self.thresholdLayer];
+
+    self.barLayers = [NSMutableArray array];
+}
+
+- (void)layoutSubviews {
+    [super layoutSubviews];
+
+    if (!CGRectEqualToRect(self.scrollLayer.frame, self.bounds)) {
+        self.scrollLayer.frame = self.bounds;
+
+        CGRect thresholdLineRect = CGRectMake(0, self.frame.size.height - [self renderDurationTargetMilliseconds], self.frame.size.width, 1);
+        self.thresholdLayer.path = CGPathCreateWithRect(thresholdLineRect, nil);
+    }
+}
+
+- (void)updatePathWithFrameDuration:(CFTimeInterval)frameDuration {
+    [CATransaction begin];
+    [CATransaction setDisableActions:YES];
+
+    self.currentX += MBXFrameTimeBarWidth;
+
+    CAShapeLayer *bar = [self barWithFrameDuration:frameDuration];
+    bar.position = CGPointMake(self.currentX, self.frame.size.height);
+
+    [self.scrollLayer addSublayer:bar];
+    [self.barLayers addObject:bar];
+
+    [self.scrollLayer scrollToPoint:CGPointMake(self.currentX - self.frame.size.width, 0)];
+
+    [self removeStaleBars];
+
+    [CATransaction commit];
+}
+
+- (CGFloat)renderDurationTargetMilliseconds {
+    CGFloat target = (1.0 / UIScreen.mainScreen.maximumFramesPerSecond) * MBXFrameTimeExaggeration;
+    return [self roundedFloat:target];
+}
+
+- (CGFloat)roundedFloat:(CGFloat)f {
+#if TARGET_OS_IPHONE || TARGET_OS_SIMULATOR
+    CGFloat scaleFactor = [UIScreen mainScreen].nativeScale;
+#elif TARGET_OS_MAC
+    CGFloat scaleFactor = [NSScreen mainScreen].backingScaleFactor;
+#endif
+    return round(f * scaleFactor) / scaleFactor;
+}
+
+- (CAShapeLayer *)barWithFrameDuration:(CFTimeInterval)frameDuration {
+    CAShapeLayer *bar = [CAShapeLayer layer];
+
+    CGRect barRect = CGRectMake(0, 0, MBXFrameTimeBarWidth, -(fminf(frameDuration * MBXFrameTimeExaggeration, self.frame.size.height)));
+    UIBezierPath *barPath = [UIBezierPath bezierPathWithRect:barRect];
+    bar.path = barPath.CGPath;
+    bar.fillColor = [self colorForFrameDuration:frameDuration].CGColor;
+
+    return bar;
+}
+
+- (UIColor *)colorForFrameDuration:(CFTimeInterval)frameDuration {
+    CGFloat renderDurationTargetMilliseconds = [self renderDurationTargetMilliseconds];
+    frameDuration *= MBXFrameTimeExaggeration;
+
+    if (frameDuration < renderDurationTargetMilliseconds && frameDuration > (renderDurationTargetMilliseconds * 0.75)) {
+        // Warning: Orange
+        return [UIColor colorWithRed:(CGFloat)(255.f/255.f) green:(CGFloat)(154.f/255.f) blue:(CGFloat)(82.f/255.f) alpha:1.f];
```

Comment:
I wonder if it's worth adding these UIColors as properties?

Reply:
Good idea, this was left over from the hackathon mindset.